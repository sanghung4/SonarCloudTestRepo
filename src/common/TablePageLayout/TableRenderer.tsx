import {
  Divider,
  Table,
  TableBody,
  TableContainer,
  SxProps,
  Theme
} from '@dialexa/reece-component-library';

import { TableInstance } from 'react-table';

import {
  tableSx,
  TableRendererBodyLoading,
  TableRendererBodyRow,
  TableRendererFooter,
  TableRendererHead
} from 'common/TablePageLayout/util';

export type TableRendererProps<TData extends object = {}> = {
  loading?: boolean;
  noResultsMessage?: string;
  resultsCount?: number;
  resultsCountText?: string;
  tableInstance: TableInstance<TData>;
  testId: string;
  hasGroups?: boolean;
  onRowClick?: (row: TData) => void;
  customHeaderProps?: SxProps<Theme>;
  customCellProps?: SxProps<Theme>;
  isWaterworks?: boolean;
  noResultsContactMessage?: string;
  noResultsContactBranch?: string;
  primaryKey: string;
};

/**
 * This expects a table instance from react table to be passed to it.
 * Requires pagination to be included.
 */
function TableTemplate<TData extends object = {}>(
  props: TableRendererProps<TData>
) {
  return (
    <>
      <TableContainer sx={{ p: '0 !important' }}>
        <Table
          {...props.tableInstance.getTableProps()}
          data-testid={props.testId}
          sx={tableSx}
        >
          <TableRendererHead {...props} />
          {props.loading || props.resultsCount === 0 ? (
            <TableRendererBodyLoading {...props} />
          ) : (
            <TableBody {...props.tableInstance.getTableBodyProps()}>
              {props.tableInstance.page.map((row, rowIndex) => {
                props.tableInstance.prepareRow(row);
                return (
                  <TableRendererBodyRow
                    {...props}
                    row={row}
                    rowIndex={rowIndex}
                    key={rowIndex}
                  />
                );
              })}
            </TableBody>
          )}
        </Table>
      </TableContainer>
      <Divider />
      <TableRendererFooter {...props} />
    </>
  );
}

export default TableTemplate;
