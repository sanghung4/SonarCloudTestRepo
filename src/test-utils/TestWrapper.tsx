import React, { JSXElementConstructor, ReactElement, ReactNode } from 'react';

import { ThemeProvider } from '@dialexa/reece-component-library';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { OktaAuth } from '@okta/okta-auth-js';
import { render as jestRender, RenderResult } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { MemoryRouter } from 'react-router-dom';

import { UserContextType, AuthContext } from 'AuthProvider';
import { Permission } from 'common/PermissionRequired';
import i18n from 'test-utils/mockI18N';
import {
  SelectedAccountsContext,
  SelectedAccountsContextType
} from 'providers/SelectedAccountsProvider';
import { CartContext, CartContextType } from 'providers/CartProvider';
import {
  CheckoutContext,
  CheckoutContextType,
  initialCheckoutContextValues
} from 'Checkout/CheckoutProvider';

/**
 * NOTE:
 * To setup mocks follow the examples here:
 * https://www.apollographql.com/docs/react/development-testing/testing/
 *
 * To get a successful response in your test (not just the loading state)
 * you can do the following:
 *
 * it('Should render', async () => {
 *   const { container } = renderMockRouter(<MyComponent />, mocks);
 *
 *   await waitFor(() => new Promise((res) => setTimeout(res, 0)));
 *
 *   expect(container).toMatchSnapshot();
 * });
 *
 */

/******************************/
/* Types                      */
/******************************/
type Props = {
  children?: ReactNode;
  authConfig?: Partial<UserContextType>;
  selectedAccountsConfig?: Partial<SelectedAccountsContextType>;
  cartConfig?: Partial<CartContextType>;
  checkoutConfig?: Partial<CheckoutContextType>;
  mocks?: ReadonlyArray<MockedResponse>;
};

type WrapperProps<TKey extends keyof Props> = {
  value: Props[TKey];
  children: React.ReactNode;
};

/******************************/
/* Defaults                   */
/******************************/
const defaultAuth: UserContextType = {
  authState: {
    isAuthenticated: true
  },
  isLoggingOut: false,
  oktaAuth: new OktaAuth({ issuer: 'http://test.com' }),
  profile: {
    userId: 'testuser',
    permissions: Object.values(Permission),
    isEmployee: false,
    isVerified: false
  }
};

const defaultSelectedAccounts: SelectedAccountsContextType = {
  clearAccounts: () => {},
  updateAccounts: () => {},
  updateShippingBranchId: () => {},
  loading: false,
  selectedAccounts: {
    billTo: { erpAccountId: '' },
    shipTo: { erpAccountId: '' },
    shippingBranchId: ''
  },
  isMincron: false,
  isEclipse: false
};

const defaultCartContextValue: CartContextType = {
  addItemToCart: () => {},
  addItemsToCart: () => {},
  addAllListItemsToCart: () => {},
  cart: undefined,
  cartLoading: false,
  checkingOutWithQuote: false,
  contractBranch: undefined,
  clearContract: () => {},
  clearQuote: () => {},
  deleteCartItems: () => new Promise(() => {}),
  disableAddToCart: false,
  deleteItem: () => {},
  getUserCart: () => new Promise(() => {}),
  itemCount: 0,
  itemLoading: undefined,
  lineNotes: {},
  releaseContractToCart: () => {},
  setContract: () => {},
  setLineNotes: () => {},
  setQuoteId: () => {},
  setQuoteData: () => {},
  setQuoteShipToId: () => {},
  setPreviousCart: () => {},
  setSelectedBranch: () => {},
  updateCart: undefined,
  updateDelivery: undefined,
  updateItemQuantity: () => {},
  updateWillCall: undefined,
  updateWillCallBranch: undefined,
  refreshCart: undefined,
  isWillCall: false,
  deleteCreditCardFromCart: () => new Promise(() => {})
};

/******************************/
/* Wrappers                   */
/******************************/
function SelectedAccountsProviderWrapper({
  value,
  children
}: WrapperProps<'selectedAccountsConfig'>) {
  if (value) {
    return (
      <SelectedAccountsContext.Provider
        value={{ ...defaultSelectedAccounts, ...value }}
      >
        {children}
      </SelectedAccountsContext.Provider>
    );
  }

  return <>{children}</>;
}

function CartProviderWrapper({ value, children }: WrapperProps<'cartConfig'>) {
  if (value) {
    return (
      <CartContext.Provider value={{ ...defaultCartContextValue, ...value }}>
        {children}
      </CartContext.Provider>
    );
  }

  return <>{children}</>;
}

function CheckoutProviderWrapper({
  value,
  children
}: WrapperProps<'checkoutConfig'>) {
  if (value) {
    return (
      <CheckoutContext.Provider
        value={{ ...initialCheckoutContextValues, ...value }}
      >
        {children}
      </CheckoutContext.Provider>
    );
  }

  return <>{children}</>;
}

function TestWrapper(props: Props) {
  return (
    <ThemeProvider>
      <I18nextProvider i18n={i18n}>
        <AuthContext.Provider value={{ ...defaultAuth, ...props.authConfig }}>
          <MockedProvider mocks={props.mocks ?? []}>
            <SelectedAccountsProviderWrapper
              value={props.selectedAccountsConfig}
            >
              <CartProviderWrapper value={props.cartConfig}>
                <CheckoutProviderWrapper value={props.checkoutConfig}>
                  {props.children}
                </CheckoutProviderWrapper>
              </CartProviderWrapper>
            </SelectedAccountsProviderWrapper>
          </MockedProvider>
        </AuthContext.Provider>
      </I18nextProvider>
    </ThemeProvider>
  );
}

function jestRenderFn(wrapped: JSX.Element) {
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined
  });
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRenderFn(wrapped);
}
export type TestRender = RenderResult<
  typeof import('@testing-library/dom/types/queries'),
  Element | DocumentFragment
>;

export function renderWithComponent(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return {
    render: jestRenderFn(wrapped),
    component: wrapped
  };
}

export default TestWrapper;
