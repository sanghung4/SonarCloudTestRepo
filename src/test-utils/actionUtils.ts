import { fireEvent, findByTestId, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export const fillFormTextInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const fillMaskedFormTextInput = async (
  testId: string,
  value: string
) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, `${testId}-input`);
    await userEvent.type(textInput, value);
  });
};

export const clickButton = async (testId: string) => {
  await act(async () => {
    const button = await findByTestId(document.body, testId);
    await fireEvent.click(button);
  });
};

export const clickRadio = async (testId: string, optionIndex: number) => {
  await act(async () => {
    const radio = await findByTestId(
      document.body,
      `${testId}-radio-${optionIndex}`
    );
    await userEvent.click(radio);
  });
};

export const fillFormSelectInput = async (testId: string, value: string) => {
  await act(async () => {
    const selectInput = await findByTestId(document.body, `${testId}-input`);
    await fireEvent.change(selectInput, { target: { value } });
    await fireEvent.keyDown(window, { key: 'Enter' });
  });
};

export const focusInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.focus(input);
  });
};

export const changeInput = async (testId: string, value: string) => {
  await act(async () => {
    const textInput = await findByTestId(document.body, testId);
    await fireEvent.change(textInput, { target: { value } });
  });
};

export const blurInput = async (testId: string) => {
  await act(async () => {
    const input = await findByTestId(document.body, testId);
    await fireEvent.blur(input);
  });
};

export const keyboardDown = async (
  key: string,
  element?: Element | Node | Document | Window
) => {
  await act(async () => {
    await fireEvent.keyDown(element ?? window, { key });
  });
};
