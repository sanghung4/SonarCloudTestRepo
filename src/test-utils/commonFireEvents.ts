import { act, fireEvent } from '@testing-library/react';

export function actClickOnElement(element: HTMLElement) {
  act(() => {
    fireEvent.click(element);
  });
}
export function actUpdateInput(input: HTMLElement, value: string) {
  act(() => updateInput(input, value));
}
export function updateInput(input: HTMLElement, value: string) {
  fireEvent.focus(input);
  fireEvent.change(input, { target: { value } });
  fireEvent.blur(input);
}
