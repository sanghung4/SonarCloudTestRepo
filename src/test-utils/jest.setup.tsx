import React, { JSXElementConstructor, ReactElement, ReactNode } from "react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { MemoryRouter } from "react-router-dom";
import { render as jestRender } from "@testing-library/react";

export const mockApollo = jest.mock(
  "@apollo/client/testing",
  () => MockedProvider
);

interface Props {
  children?: ReactNode;
  mocks?: ReadonlyArray<MockedResponse>;
}

function TestWrapper(props: Props) {
  return (
    <MockedProvider mocks={props.mocks ?? []}>{props.children}</MockedProvider>
  );
}

export function render(component: ReactNode, config?: Props) {
  const wrapped = <TestWrapper {...config}>{component}</TestWrapper>;
  return jestRender(wrapped, {
    wrapper: MemoryRouter as
      | JSXElementConstructor<{
          children: ReactElement<any, string | JSXElementConstructor<any>>;
        }>
      | undefined,
  });
}
