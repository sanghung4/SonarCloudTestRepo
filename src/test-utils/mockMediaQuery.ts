import mediaQuery from 'css-mediaquery';

function createMatchMedia(width: number) {
  return (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: () => jest.fn(),
    removeListener: () => jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn()
  });
}

export const breakpoints = {
  mobile: 414,
  desktop: 1200
};

export const setBreakpoint = (breakpoint: 'mobile' | 'desktop' = 'desktop') => {
  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: breakpoints[breakpoint]
  });
  window.dispatchEvent(new Event('resize'));

  window.matchMedia = createMatchMedia(window.innerWidth);
};
