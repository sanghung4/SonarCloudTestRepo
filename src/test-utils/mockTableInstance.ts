import {
  Cell,
  IdType,
  Row,
  TableInstance,
  UseRowStateLocalState
} from 'react-table';

export function mockRow<T extends object>(
  values: Record<IdType<T>, any>,
  original: T,
  state: UseRowStateLocalState<T, unknown>
): Row<T> {
  return {
    allCells: [],
    cells: [],
    values,
    getRowProps: jest.fn(),
    index: 0,
    original,
    id: '',
    subRows: [],
    isExpanded: false,
    canExpand: false,
    toggleRowExpanded: jest.fn(),
    getToggleRowExpandedProps: jest.fn(),
    depth: 0,
    isGrouped: false,
    groupByID: '',
    groupByVal: '',
    leafRows: [],
    isSelected: false,
    isSomeSelected: false,
    toggleRowSelected: jest.fn(),
    getToggleRowSelectedProps: jest.fn(),
    state,
    setState: jest.fn()
  } as Row<T>;
}

export function mockCell<T extends object>(
  value: T,
  column: IdType<T>,
  state: UseRowStateLocalState<T, unknown>
) {
  const cell: Cell<T> = {
    column: {
      id: column,
      isVisible: false,
      render: jest.fn(),
      totalLeft: 0,
      totalWidth: 0,
      getHeaderProps: jest.fn(),
      getFooterProps: jest.fn(),
      toggleHidden: jest.fn(),
      getToggleHiddenProps: jest.fn(),
      depth: 0,
      canFilter: false,
      setFilter: jest.fn(),
      filterValue: '',
      preFilteredRows: [],
      filteredRows: [],
      canGroupBy: false,
      isGrouped: false,
      groupedIndex: 0,
      toggleGroupBy: jest.fn(),
      getGroupByToggleProps: jest.fn(),
      getResizerProps: jest.fn(),
      canResize: false,
      isResizing: false,
      canSort: false,
      toggleSortBy: jest.fn(),
      getSortByToggleProps: jest.fn(),
      clearSortBy: jest.fn(),
      isSorted: false,
      sortedIndex: 0,
      isSortedDesc: false
    },
    row: mockRow(
      value as Record<IdType<T>, any>,
      value,
      {} as UseRowStateLocalState<T, unknown>
    ),
    value,
    getCellProps: jest.fn(),
    render: jest.fn(),
    isGrouped: false,
    isPlaceholder: false,
    isAggregated: false,
    state,
    setState: jest.fn()
  };
  return cell;
}

export const mockTableInstance: TableInstance = {
  state: {
    columnOrder: [],
    expanded: {},
    filters: [],
    globalFilter: null,
    groupBy: [],
    pageSize: 0,
    pageIndex: 0,
    columnResizing: {
      columnWidth: 0,
      columnWidths: 0,
      headerIdWidths: {}
    },
    selectedRowIds: {},
    rowState: {},
    sortBy: []
  },
  plugins: [],
  dispatch: jest.fn(),
  columns: [],
  allColumns: [],
  visibleColumns: [],
  headerGroups: [],
  footerGroups: [],
  headers: [],
  flatHeaders: [],
  rows: [],
  rowsById: {},
  getTableProps: jest.fn(),
  getTableBodyProps: jest.fn(),
  prepareRow: jest.fn(),
  flatRows: [],
  totalColumnsWidth: 0,
  allColumnsHidden: false,
  toggleHideColumn: jest.fn(),
  setHiddenColumns: jest.fn(),
  toggleHideAllColumns: jest.fn(),
  getToggleHideAllColumnsProps: jest.fn(),
  getHooks: jest.fn(),
  setColumnOrder: jest.fn(),
  preExpandedRows: [],
  expandedRows: [],
  expandedDepth: 0,
  isAllRowsExpanded: false,
  toggleRowExpanded: jest.fn(),
  toggleAllRowsExpanded: jest.fn(),
  preFilteredRows: [],
  preFilteredFlatRows: [],
  preFilteredRowsById: {},
  filteredRows: [],
  filteredFlatRows: [],
  filteredRowsById: {},
  setFilter: jest.fn(),
  setAllFilters: jest.fn(),
  preGlobalFilteredRows: [],
  preGlobalFilteredFlatRows: [],
  preGlobalFilteredRowsById: {},
  globalFilteredRows: [],
  globalFilteredFlatRows: [],
  globalFilteredRowsById: {},
  setGlobalFilter: jest.fn(),
  preGroupedRows: [],
  preGroupedFlatRows: [],
  preGroupedRowsById: {},
  groupedRows: [],
  groupedFlatRows: [],
  groupedRowsById: {},
  onlyGroupedFlatRows: [],
  onlyGroupedRowsById: {},
  nonGroupedFlatRows: [],
  nonGroupedRowsById: {},
  toggleGroupBy: jest.fn(),
  page: [],
  pageCount: 0,
  pageOptions: [],
  canPreviousPage: false,
  canNextPage: false,
  gotoPage: jest.fn(),
  previousPage: jest.fn(),
  nextPage: jest.fn(),
  setPageSize: jest.fn(),
  toggleRowSelected: jest.fn(),
  toggleAllRowsSelected: jest.fn(),
  getToggleAllRowsSelectedProps: jest.fn(),
  getToggleAllPageRowsSelectedProps: jest.fn(),
  isAllRowsSelected: false,
  selectedFlatRows: [],
  setRowState: jest.fn(),
  setCellState: jest.fn(),
  preSortedRows: [],
  setSortBy: jest.fn(),
  toggleSortBy: jest.fn()
};
