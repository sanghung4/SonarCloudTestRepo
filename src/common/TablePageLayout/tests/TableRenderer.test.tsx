import TableRenderer, {
  TableRendererProps
} from 'common/TablePageLayout/TableRenderer';
import { setBreakpoint } from 'test-utils/mockMediaQuery';
import { mockTableInstance } from 'test-utils/mockTableInstance';
import { render } from 'test-utils/TestWrapper';

const mockProps: TableRendererProps = {
  tableInstance: mockTableInstance,
  testId: 'test',
  primaryKey: 'key'
};

describe('common - TablePageLayout - TableRenderer', () => {
  it('Expect to match snapshot with basic props on desktop', () => {
    setBreakpoint('desktop');
    const { container } = render(<TableRenderer {...mockProps} />);
    expect(container).toMatchSnapshot();
  });
  it('Expect to match snapshot with basic props on mobile', () => {
    setBreakpoint('mobile');
    const { container } = render(<TableRenderer {...mockProps} />);
    expect(container).toMatchSnapshot();
  });
});
