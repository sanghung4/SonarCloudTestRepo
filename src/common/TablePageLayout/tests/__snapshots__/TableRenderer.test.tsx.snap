// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common - TablePageLayout - TableRenderer Expect to match snapshot with basic props on desktop 1`] = `
<div>
  <div
    class="MuiTableContainer-root css-8cb3ox-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-qzyhip-MuiTable-root"
      data-testid="test"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      />
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      />
    </table>
  </div>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-di86mu-MuiDivider-root"
  />
  <div
    class="MuiBox-root css-11ezxrp"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-1vbz8i-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-auto css-kqwn4m-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-19z5v8g"
          data-testid="total-undefined-count"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`common - TablePageLayout - TableRenderer Expect to match snapshot with basic props on mobile 1`] = `
<div>
  <div
    class="MuiTableContainer-root css-8cb3ox-MuiTableContainer-root"
  >
    <table
      class="MuiTable-root css-qzyhip-MuiTable-root"
      data-testid="test"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      />
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      />
    </table>
  </div>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-di86mu-MuiDivider-root"
  />
  <div
    class="MuiBox-root css-11ezxrp"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column-reverse MuiGrid-grid-xs-12 css-ykx38k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-auto css-kqwn4m-MuiGrid-root"
      >
        <div
          class="MuiBox-root css-19z5v8g"
          data-testid="total-undefined-count"
        />
      </div>
    </div>
  </div>
</div>
`;
